documentation/optee_design.md

```diff
@@ -198,6 +198,85 @@ that should be called when Trusted OS receives standard or fast calls, FIQ,
 
+## Synchronization
+OP-TEE has three primitives for synchornization of threads and CPUs,
+spin-lock, mutex, and condvar.
+
+### Spin-lock
+A spin-lock is represented as an `unsigned int`. This is the most primitive
+lock. Interrupts should be disabled before attempting to take a spin-lock
+and should remain disabled until the lock is released.
+
+A spin-lock is initialized with `SPINLOCK_UNLOCK`.
+
+`cpu_spin_lock()` locks a spin-lock
+
+`cpu_spin_trylock()` locks a spin-lock if unlocked and returns `0` else if
+the spin-spin else the spin-lock is unchanged and the function returns `!0`.
+
+`cpu_spin_unlock()` unlocks a spin-lock
+
+### Mutex
+A mutex is represented by `struct mutex`. A mutex can be locked and
+unlocked with interrupts enabled or disabled, but only from a normal
+thread. A mutex can't be used in an interrupt handler, abort handler or
+before a thread has been selected for the CPU.
+
+A mutex is initialized with either `MUTEX_INITIALIZER` or `mutex_init()`.
+
+`mutex_lock()` locks a mutex. If the mutex is unlocked this is a fast
+operation, else the function issues an RPC to wait in normal world.
+
+`mutex_unlock()` unlocks a mutex. If there's no waiters this is a fast
+operation, else the function issues an RPC to wake up a waiter in normal
+world.
+
+`mutex_trylock()` locks a mutex if unlocked and returns `true` else the
+mutex is unchanged and the function returns `false`.
+
+`mutex_destroy()` asserts that the mutex is unlocked and there's no
+waiters, after this the memory used by the mutex can be freed.
+
+When a mutex is locked it's own by the thread calling `mutex_lock()` or
+`mutex_trylock()`, the mutex may only be unlocked by the thread owning the
+mutex.
+
+A thread should not exit to TA user space when holding a mutex, the only
+exception is the *big lock* mutex which is handled in a special way.
+
+### Condvar
+A condvar is represented by `struct condvar`. A condvar is similar to a
+pthread_condvar_t in the pthreads standard, only less advanced. Condition
+variables are used to wait for some condition to be fulfilled and are
+always used together a mutex. Once a condition variable has been used with
+a certain mutex, it must only be used with that mutex until destroyed.
+
+A condvar is initialized with `CONDVAR_INITIALIZER` or `condvar_init()`.
+
+`condvar_wait()` atomically unlocks the supplied mutex waits in normal
+world via an RPC for the condition variable to be signaled, when the
+function returns the mutex is locked again.
+
+`condvar_signal()` wakes up one waiter of the condition variable (waiting
+in `condvar_wait()`
+
+`condvar_broadcast()` wake up all waiters of the condition variable.
+
+The caller of `condvar_signal()` or `condvar_broadcast()` should hold the
+mutex associated with the condition variable to guarantee that a waiter
+doesn't miss the signal.
+
+## Big lock
+The *big lock* mutex protects secure world against unsupported concurrent
```

Comment:
Would be worth to know what is "unsupported". Right now, we have only 1 thread active in the trusted os. But when leaving the trusted os (rpc or normal exit), then we're able to take the execution of another TA.